integration-tests/vault/src/test/java/io/quarkus/vault/VaultTransitITCase.java

```diff
@@ -314,4 +317,114 @@ public void adminKey() {
         assertNull(transitSecretEngine.readKey(KEY_NAME));
     }
 
+    @Test
+    public void asymmetricReadECDSAKey() {
+
+        assertFalse(transitSecretEngine.listKeys().contains(KEY_NAME));
+        transitSecretEngine.createKey(KEY_NAME, new KeyCreationRequestDetail().setType("ecdsa-p256"));
+        assertTrue(transitSecretEngine.listKeys().contains(KEY_NAME));
+
+        VaultTransitKeyDetail mykey = transitSecretEngine.readKey(KEY_NAME);
+        assertTrue(mykey instanceof VaultTransitAsymmetricKeyDetail);
+        assertEquals(KEY_NAME, mykey.getName());
+        assertFalse(mykey.isExportable());
+        assertFalse(mykey.isDeletionAllowed());
+        assertFalse(mykey.isSupportsDecryption());
+        assertFalse(mykey.isSupportsEncryption());
+        assertFalse(mykey.isSupportsDerivation());
+        assertTrue(mykey.isSupportsSigning());
+        assertEquals(mykey.getType(), "ecdsa-p256");
+        assertEquals(1, mykey.getKeys().size());
+        assertTrue(mykey.getKeys().containsKey("1"));
+        assertEquals(1, mykey.getVersions().size());
+        assertTrue(mykey.getVersions().containsKey("1"));
+        assertNotNull(mykey.getVersions().get("1").getCreationTime());
+        assertTrue(mykey.getVersions().get("1") instanceof VaultTransitAsymmetricKeyVersion);
+        assertNotNull(((VaultTransitAsymmetricKeyVersion) mykey.getVersions().get("1")).getPublicKey());
```

Comment:
This does not look good IMHO

Reply:
There could be only symmetric and asymmetric keys - so a more user friendly approach is to provide an option to retrieve symmetric keys and asymmetric keys as opposed to expecting the users to cast

Reply:
This was for testing. The new detail and version objects make it easy to test and access versions without excessive casting.

For example...
```java
if (mykey instanceof VaultTransitAsymmetricKeyDetail) {
  VaultTransitAsymmetricKeyDetail myAsymKey = (VaultTransitAsymmetricKeyDetail)mykey;
  // myAsymKey.versions is typed to asymmetric versions
  for (VaultTransitAsymmetricKeyVersion version : myAsymKey.versions) {
    ... 
  }
} else {
  VaultTransitSymmetricKeyDetail mySymKey = (VaultTransitSymmetricKeyDetail)mykey;
  // mySymKey.versions is typed to symmetric versions
  for (VaultTransitSymmetricKeyVersion version : mySymKey.versions) {
    ... 
  }
}
```

With Java 14... it's even easier
```java
if (mykey instanceof VaultTransitAsymmetricKeyDetail myAsymKey) {
  for (VaultTransitAsymmetricKeyVersion version : myAsymKey.versions) {
    ... 
  }
}
```

Reply:
To be clear all versions in a detail object will be of the same type. The Vault API will not allow mixing asymmetric and symmetric key versions. So, with your suggestion, you'd have two separate maps of versions and one would always be empty. You'd have to check the size of the maps to see which one is not empty and use that one. That's more work and more confusing than checking the detail object's type and casting the detail object to get the correctly typed versions.

Reply:
I'm sorry but it does not look easy to me. This is exactly why I doubted this approach in the issue discussion `instranceof plus cast` - it is not really how I'd expect Quarkus Vault users accessing the keys in the new version of this API

Reply:
🤷‍♂️ That's a problem you seem to have with Java.

Again, your suggestion appears to be to create _**two**_ maps of versions, one for symmetric versions and one or asymmetric versions. That is **not** how the data is organized because each `VaultTransitKeyDetail` can only contain one type of versions. You'd have two maps and one would always be completely empty. In this scenario, without using`instanceof` how would a user check if it's an asymmetric or symmetric key?

Reply:
Perhaps a single map can be used but in addition to the users being able to access this map they can have options for getting specific representations ? How about adding a few more helpers to get a list of asymmetric/symmetric keys ? That would be a reasonable compromise IMHO

Reply:
Also, I was thinking yesterday, should the users have access to `PublicKey` instances as opposed to them having to parse PEM content ? May be both options should be available...

Reply:
> Perhaps a single map can be used but in addition to the users being able to access this map they can have options for getting specific representations ?

The user must still first determine if the key is symmetric or asymmetric prior to asking for specific representations. If the user does that via `instanceof`, then they have access to the specific (and type correct) representations via the detail subclasses.

It might help to reexamine the classes `VaultTransitKeyDetail` and it's subclasses `VaultTransitAsymmetricKeyDetail` and `VaultTransitSymmetricKeyDetail`.  Once a user has determined wether a key detail object is symmetric or asymmetric (I'm using `instanceof` to determine this), a single cast gives users exactly what you are asking for, specific representations of the versions.

Reply:
> Also, I was thinking yesterday, should the users have access to `PublicKey` instances as opposed to them having to parse PEM content ? May be both options should be available...

I like this idea but there's a bit of a problem.  There's no good PEM parser in Java security so to support this we'd have to bring in BouncyCastle or roll our own PEM parser.

Also, `VaultTransitKeyExportDetail` represents its PEM private key versions as `String`s. So, currently there is consistency.

Reply:
I don't agree with the `instanceof/cast` approach - IMHO it is not user friendly at all - but I'll let Vincent decide.

I don't think a 3rd party parser is needed but agreed representing keys are strings is fine for now

Reply:
catching up on the discussions.

I am not sure the `VaultTransitKeyDetail` is bringing a lot of value with the 2 subclasses. or at the very least the abstract superclass should introduce field `Map<String, T> versions`, with `VaultTransitKeyDetail` being parameterized with `T extends VaultTransitKeyDetail`. then subclasses should defined `T` consistently: `VaultTransitSymmetricKeyVersion` for `VaultTransitSymmetricKeyDetail`, and `VaultTransitAsymmetricKeyVersion` for `VaultTransitAsymmetricKeyDetail`.

the other option is to keep a single class `VaultTransitKeyDetail` like before , and add field `Map<String, ? extends VaultTransitKeyVersion> versions`. I have a slight preference for that solution, since after you do work described just above, you will end up with empty `VaultTransitSymmetricKeyDetail` and `VaultTransitAsymmetricKeyDetail` classes, serving essentially as markers to get the appropriate parametrized types for the `versions` map.

what do you all think?

Reply:
First, we need "some" method for users to check if it's a symmetric or asymmetric key. We should _not_ require them to use the `type` property because as Vault adds more key types, the `type` field domain will grow and become a maintenance headache. So, we need a simple check for asymmetric versus symmetric.

We can do that with subclasses (as I have done) or we can do it with another explicit field (e.g. `kind` or `isAsymmetric`). The method I have implemented using subclasses has advantages in that both Java 14+ and Kotlin have pattern matching for this exact scenario.  In Java 14+ `instanceof` checks can easily elide the cast and Kotlin automatically casts the checked values. With an explicit field (e.g. `kind`) the user would be required to check the field and then perform an explicit cast anyway in both languages.

##### Examples for `instanceof` method in Java & Kotlin

`instanceof` (Java 14+)
```java
if (mykey instanceof VaultTransitAsymmetricKeyDetail myAsymKey) {
  for (var version : myAsymKey.versions) {
    ... 
  }
}
```

`is` (Kotlin)
```kotlin
if (mykey is VaultTransitAsymmetricKeyDetail) {
  for (version in mykey.versions) {
    ... 
  }
}
```

##### Example for explicit field method in Java  & Kotlin

```java
if (mykey.kind == VaultTransitKeyDetail.Kind.Asymmetric) {
  var myAsymKey = (VaultTransitAsymmetricKeyDetail) mykey;
  for (var version : myAsymKey.versions) {
    ...
  }
}
```

```kotlin
if (mykey.kind == VaultTransitKeyDetail.Kind.Asymmetric) {
  var myAsymKey = mykey as VaultTransitAsymmetricKeyDetail;
  for (version in myAsymKey.versions) {
    ...
  }
}
```

Reply:
One final argument for subclasses, they're more future proof.  As Vault API grows and changes we can now distinctly add properties to either symmetric or asymmetric keys and their versions.

Reply:
I agree with using a hierarchy. but I am not convinced that we should have it on the detail, plus the version. I am thinking that having it on the version is enough, which follows vault's api where the distinction is only made in the `keys` field.
could we get away with a single class `VaultTransitKeyDetail`, and the 3 classes for the version: `VaultTransitKeyVersion`, `VaultTransitAsymmetricKeyVersion` and `VaultTransitSymmetricKeyVersion`?
then the `instanceof` is only happening if you manipulate asymmetric keys, since for symmetric you can get the creation time from the base class anyway.
or do you think that it is important to make this distinction in the detail already?

Reply:
I like having the detail be distinct for two reasons. First, you do an `instanceof` check and cast once on the detail and can now assume all the version objects are the same and can be used without casting; alternatively you'll cast each version in a loop. Second, it's future proofing for other areas where symmetric and asymmetric keys may differ in the future. Vault Transit is new and changing rapidly (as evidenced by these changes) so we could be required to break them out in the future regardless.

To your last point, remember if you only want the `creationTime`, `VaultTransitKeyDetail` users can get that in the current implementation without checking or downcasting because it's `versions` property is `Map<String, ? extends VaultTransitKeyVersion>`.

Reply:
> First, you do an instanceof check and cast once on the detail and can now assume all the version objects are the same and can be used without casting

agreed. but that is the only value added by those subclasses, as they will be empty once you start using generics.

> alternatively you'll cast each version in a loop

only if you know it is asymmetric, and you need the extra information (beside the creation time), of if you do not know the type (and usually you do I suspect).

> Second, it's future proofing for other areas where symmetric and asymmetric keys may differ in the future

if they differ later, we will always be able to refactor them into the hierarchy you are proposing today.

to progress on this particular point, could you start parametrizing the detail base class with generics? this should empty out the subclasses, and help reason later about whether or not we want to drop the detail hirerachy altogether. thanks.

Reply:
I gentrified `VaultTransitKeyDetail` to include the version class as a type parameter. 

A nice side-effect of this is that it announces to users that each version entry is always of the same exact subtype.